lib/commonio.c

```diff
@@ -110,9 +110,9 @@ static int check_link_count (const char *file, bool log)
 
 	if (sb.st_nlink != 2) {
 		if (log) {
-			(void) fprintf (shadow_logfd,
-			                "%s: %s: lock file already used (nlink: %u)\n",
-			                shadow_progname, file, sb.st_nlink);
+			fprintf(shadow_logfd,
+			        "%s: %s: lock file already used (nlink: %ju)\n",
+			        shadow_progname, file, (uintmax_t) sb.st_nlink);
```

Comment:
Can you give more details as to why you are changing this?

Reply:
`st.nlink` is of type `nlink_t`.  `%u` is for printing `unsigned int`, not `nlink_t`.

Most typedefs don't have a printf(3) length modifier, with few exceptions, such as `z` for `[s]size_t`, `t` for `ptrdiff_t`, and `j` for `[u]intmax_t`.

So, for printing typedefs, since we don't know the underlying type (at least not portably), the best thing is to convert to a widest type (usually `[unsigned] long long` or `[u]intmax_t`, and use its length modifier `ll` or `j`, respectively).

This is documented in system_data_types(7) (although I need to rewrite that page after having split it into many smaller pages):

```
     Most of the integer types described in this  page  don’t  have  a
     corresponding  length modifier for the printf(3) and the scanf(3)
     families of functions.  To print a value of an integer type  that
     doesn’t  have  a  length modifier, it should be converted to int‐
     max_t or uintmax_t by an explicit cast.  To scan into a  variable
     of an integer type that doesn’t have a length modifier, an inter‐
     mediate  temporary  variable of type intmax_t or uintmax_t should
     be used.  When copying from the temporary variable to the  desti‐
     nation variable, the value could overflow.  If the type has upper
     and  lower limits, the user should check that the value is within
     those limits, before actually copying the value.  The example be‐
     low shows how these conversions should be done.
```